### ISSUE TITLE
Content Security Policy: The page's settings blocked the loading of a resource at self

### ISSUE THREAD

**Issue by zubanst** [owncloud/music#498]
OC 8.1.1, Music 0.3.8, FF 46.0 on Ubuntu 14.04; no sound paying music with the following error in the browser console:
Content Security Policy: The page's settings blocked the loading of a resource at self ("script-src https://www.pennyitsupport.eu 'unsafe-eval'").

Not clear though, if the missing sound is due to this error.

**Comment by ocelotsloth** [comment]
Interesting...the header set on my browser has `media` set to `"self"`.

It needs to be modified to add `data:` as well, since `"self"` doesn't cover the data URI generated by the addon.

**Comment by ocelotsloth** [comment]
And wow this is an old issue. I'll see if I can get a fix written.

**Comment by zubanst** [comment]
Well, if there is a fix, any fix is welcome as long as it makes the code more robust, though with newer versions of FF, it cannot be reproduced.

**Comment by ocelotsloth** [comment]
Unless I'm mistaken this appears to be a class expressly for this purpose?

https://github.com/owncloud/core/blob/f10d105ea2f2d0822693313ddcd6f9b2eaac6ebe/lib/public/AppFramework/Http/ContentSecurityPolicy.php

**Comment by ocelotsloth** [comment]
I suppose to be more specific, this is what is needed:
https://github.com/owncloud/core/blob/37177dd54d67443cacc0061c17f95070442b93cf/tests/lib/AppFramework/Http/ResponseTest.php#L70-L81

```php
	public function testOverwriteCsp() {
		$expected = [
			'Content-Security-Policy' => "default-src 'none';script-src 'self' 'unsafe-inline' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self';font-src 'self';connect-src 'self';media-src 'self'",
		];
		$policy = new ContentSecurityPolicy();
		$policy->allowInlineScript(true);
		$this->childResponse->setContentSecurityPolicy($policy);
		$headers = $this->childResponse->getHeaders();
		$this->assertEquals(array_merge($expected, $headers), $headers);
	}
```

**Comment by paulijar** [comment]
I have never seen any Content Security Policy errors on Firefox but Chrome seems to print one to its console whenever I play a track. The track plays fine, though, so the problem is quite trivial. But if you can contribute a neat fix, that's greatly appreciated.

**Comment by ocelotsloth** [comment]
Hmm...I can confirm that it works on Firefox, but it fails to play the song using Chromium on Arch Linux.

The CSP error is the the only error logged to the browser console with Chromium, so I assumed that it was failing for that reason. Is there something else that could be stopping the audio?

**Comment by ocelotsloth** [comment]
That said, I do see the Content Security Policy Errors on Firefox as well.

![](https://i.imgur.com/BGDSvuL.png)

**Comment by paulijar** [comment]
@ocelotsloth The difference of Chromium compared to Firefox and Chrome is that in (some installations of) Chromium there is no built-in audio codecs. On such browsers, the Music app falls back to using the aurora.js library for playback as it does not require browser support. Otherwise the app uses the SoundManager2 library (sm2). You can see the used playback engine from the console debug print `Using <playback engine> for <mimetype> URL <url>`.

It's possible that the Content Security Policy issue is a bigger problem for aurora than for sm2. At least this would explain the perceived behavior.

Also, the aurora.js library has a known bug that causes playback of some MP3 files to fail. But most files should play fine.

**Comment by paulijar** [comment]
The Content Security Policy issue has been fixed with commit 5c310e1. At least I no longer see CSP errors on Chrome. The fix shall be included in the next release which will likely be v0.5.5.

**Comment by paulijar** [comment]
The fix has now been released in the v0.5.5.
